src/plugins/plugin.legend.js

```diff
@@ -47,17 +47,36 @@ defaults._set('global', {
 			generateLabels: function(chart) {
 				var data = chart.data;
 				return helpers.isArray(data.datasets) ? data.datasets.map(function(dataset, i) {
+					var type = chart.getDatasetMeta(i).type;
+					var options = chart.options;
+					var legendOpts = options.legend;
+					var valueAtIndexOrDefault = helpers.valueAtIndexOrDefault;
+					var valueOrDefault = helpers.valueOrDefault;
+
+					var usePointStyle = legendOpts && legendOpts.labels && legendOpts.labels.usePointStyle;
+					var backgroundColor = usePointStyle ? valueAtIndexOrDefault(dataset.pointBackgroundColor, 0, dataset.backgroundColor) : dataset.backgroundColor;
+					var borderWidth = usePointStyle ? valueAtIndexOrDefault(dataset.pointBorderWidth, 0, dataset.borderWidth) : dataset.borderWidth;
+					var borderColor = usePointStyle ? valueAtIndexOrDefault(dataset.pointBorderColor, 0, dataset.borderColor) : dataset.borderColor;
+
+					var useLineStyles = (type === 'line' || type === 'radar') && !usePointStyle;
+					var borderCapStyle = useLineStyles ? dataset.borderCapStyle : 'butt';
+					var borderDash = useLineStyles ? dataset.borderDash : [];
+					var borderDashOffset = useLineStyles ? dataset.borderDashOffset : 0;
+					var borderJoinStyle = useLineStyles ? dataset.borderJoinStyle : 'miter';
+
+					var elementOpts = options.elements[useLineStyles ? 'line' : type === 'bar' ? 'rectangle' : 'point'];
+
 					return {
 						text: dataset.label,
-						fillStyle: (!helpers.isArray(dataset.backgroundColor) ? dataset.backgroundColor : dataset.backgroundColor[0]),
+						fillStyle: valueAtIndexOrDefault(backgroundColor, 0, elementOpts.backgroundColor),
```

Comment:
Are we trying to prefer `resolve` to `valueAtIndexOrDefault` now after https://github.com/chartjs/Chart.js/pull/5965? Or I think at least we want to make those variables global instead of in the function?

Reply:
Thanks for reminding me. Since this involves the checks for `usePointStyle`, I would keep using `valueAtIndexOrDefault` and just make it global.